agentendpoint/agentendpoint.go

```diff
@@ -373,16 +370,27 @@ func (c *Client) WaitForTaskNotification(ctx context.Context) {
 					c.Close()
 					return
 				}
-				clog.Errorf(ctx, "Error waiting for task: %v", err)
-				sleep := 5 * time.Second
+
 				if errors.Is(err, errResourceExhausted) {
-					sleep = retryutil.RetrySleep(resourceExhausted, 5)
 					resourceExhausted++
+					sleep = retryutil.RetrySleep(resourceExhausted, 5)
 				} else {
-					resourceExhausted = 1
+					// Retry any other errors with a modest backoff. Only retry up to 10
+					// times, at that point return, the client will be recreated during the next
+					// cycle.
+					errs++
+					clog.Warningf(ctx, "Error waiting for task (attempt %d of 10): %v", errs, err)
+					resourceExhausted = 0
+					if errs > 10 {
+						c.Close()
+						return
+					}
+					sleep = retryutil.RetrySleep(errs, 0)
 				}
 				time.Sleep(sleep)
+				continue
 			}
+			errs = 0
```

Comment:
Just an observation: it seems a little odd that this function is designed to wait forever for a task when things work, retry a few times to overcome intermittent failures, but return and let the caller retry for other failures. This manifests itself as having a few variables that track different errors and get reset in different scenarios. Would it make sense to consolidate the API retry logic for both short-term and long-term failures in one place - possibly separating retry logic for different types of errors?

Reply:
The current logic is more "continue forever unless there is a fatal error" all the errors we stop on are thought of as fatal and continuing to retry connections not super useful. The fact that main will recreate the connection is not a retry, more how the connection logic is implemented. 

I could rewrite this to never return and these "fatal" errors just force us into a longer retry loop. That could make this function make more sense (only way to stop it is to call Close() or cancel the context). It wouldn't change any logic elsewhere though as all that is till required for other situations.